compiler-plugin/src/main/kotlin/it/unibo/collektive/frontend/visitors/ImproperConstructVisitor.kt

```diff
@@ -0,0 +1,156 @@
+import org.jetbrains.kotlin.fir.expressions.FirAnonymousFunctionExpression
+import org.jetbrains.kotlin.fir.expressions.FirExpression
+import org.jetbrains.kotlin.fir.expressions.FirFunctionCall
+import org.jetbrains.kotlin.fir.expressions.FirReturnExpression
+import org.jetbrains.kotlin.fir.references.FirResolvedNamedReference
+import org.jetbrains.kotlin.fir.symbols.FirBasedSymbol
+import org.jetbrains.kotlin.fir.visitors.FirVisitorVoid
+
+/**
+ * Visitor that checks if an `evolve` or `evolving` construct can be replaced with a `share` construct.
+ */
+class ImproperConstructVisitor(
+    private val constructNameFQName: String,
+) : FirVisitorVoid() {
+    private var nestingLevel = 0
+    private var isReplaceable = false
+    private var markExpression = false
+    private var isInsideNeighboring = false
+    private var neighboringExpression: FirExpression? = null
+    private var parametersDeclarations = emptyList<FirBasedSymbol<*>>()
+    private var tracedDependentSymbols = emptyList<FirBasedSymbol<*>>()
+
+    /**
+     * Simple visitor that returns the explicit receiver expression of a `yielding` call, or `null` if it is not found.
+     */
+    private class YieldingReceiverVisitor : FirVisitorVoid() {
+        private var returnExpression: FirExpression? = null
+
+        override fun visitElement(element: FirElement) {
+            element.acceptChildren(this)
+        }
+
+        override fun visitFunctionCall(functionCall: FirFunctionCall) {
+            if (functionCall.fqName() == YIELDING_FUNCTION_FQ_NAME) {
+                returnExpression = functionCall.explicitReceiver
+            }
+        }
+
+        /**
+         * Returns the yielding receiver of a return statement, or `null` if it is not found.
+         *
+         * Example:
+         * ```kotlin
+         * return value.max(0).yielding { "example" }
+         * ```
+         * In this case, the yielding receiver is `value.max(0)`.
+         */
+        fun FirReturnExpression.getYieldingReceiver(): FirExpression? {
+            visitElement(this)
+            return returnExpression
+        }
+    }
+
+    override fun visitElement(element: FirElement) {
+        element.acceptChildren(this)
+    }
+
+    override fun visitProperty(property: FirProperty) {
+        super.visitProperty(property)
+        // trace the symbols that depends on the parameters of the construct
+        if (markExpression) {
+            tracedDependentSymbols += property.symbol
+            markExpression = false
+        }
+    }
+
+    override fun visitResolvedNamedReference(resolvedNamedReference: FirResolvedNamedReference) {
+        // mark the expression if it uses the parameters of the construct or the traced dependent symbols
+        if (resolvedNamedReference.resolvedSymbol in parametersDeclarations ||
+            resolvedNamedReference.resolvedSymbol in tracedDependentSymbols
+        ) {
+            markExpression = true
```

Comment:
Why not simply replace the assignment with the if condition? Can it cause problems?